File: 2024/02-funciones-y-alcance/solution.ts
// Ejemplos de funciones básicas
function funcionSinParametrosNiRetorno(): void {
  console.log("¡Hola desde la función sin parámetros ni retorno!");
}

function funcionConParametros(parametro1: any, parametro2: any): void {
  console.log("Parámetro 1:", parametro1);
  console.log("Parámetro 2:", parametro2);
}

function funcionConRetorno(num1: number, num2: number): number {
  return num1 + num2;
}

// Funciones dentro de funciones
function funcionExterna(): void {
  console.log("Función externa");

  function funcionInterna(): void {
    console.log("Función interna");
  }

  funcionInterna();
}

// Variable GLOBAL y LOCAL
let variableGlobal: string = "Soy global";

function funcionConVariables(): void {
  let variableLocal: string = "Soy local";
  console.log(variableGlobal);
  console.log(variableLocal);
}

// Utilizar función ya creada en el lenguaje
let numeros: number[] = [1, 2, 3, 4, 5];
let cuadrados: number[] = numeros.map((numero) => numero * numero);

console.log("Cuadrados:", cuadrados);

// Función Extra (DIFICULTAD EXTRA)
function funcionExtra(texto1: string, texto2: string): number {
  let contador: number = 0;

  for (let i: number = 1; i <= 100; i++) {
    if (i % 3 === 0 && i % 5 === 0) {
      console.log(texto1 + texto2);
    } else if (i % 3 === 0) {
      console.log(texto1);
    } else if (i % 5 === 0) {
      console.log(texto2);
    } else {
      console.log(i);
    }

    contador++;
  }

  return contador;
}

console.log("Número de impresiones:", funcionExtra("Fizz", "Buzz"));

File: 2024/02-funciones-y-alcance/solution.cs
class FuncionesAlcance
{
  // Variable global
  static int globalVariable = 10;

  public static void Execute()
  {
    // 1. Función sin parámetros ni retorno
    PrintHello();

    // 2. Función con un parámetro
    PrintMessage("Hello, C#!");

    // 3. Función con varios parámetros
    AddNumbers(5, 10);

    // 4. Función con retorno
    int result = MultiplyNumbers(3, 4);
    Console.WriteLine($"Resultado de la multiplicación: {result}");

    // 5. Funciones dentro de funciones (con función local)
    OuterFunction();

    // 6. Variables locales y globales
    int localVariable = 5;
    Console.WriteLine($"Variable local: {localVariable}");
    Console.WriteLine($"Variable global antes de modificar: {globalVariable}");
    ModifyGlobalVariable();
    Console.WriteLine($"Variable global después de modificar: {globalVariable}");

    // 7. Función con dos cadenas de texto y retorno de un número (Dificultad Extra)
    int count = PrintNumbersWithText("Fizz", "Buzz");
    Console.WriteLine($"El número de veces que se imprimió un número es: {count}");
  }

  // Función sin parámetros ni retorno
  public static void PrintHello()
  {
    Console.WriteLine("Hello, World!");
  }

  // Función con un parámetro
  public static void PrintMessage(string message)
  {
    Console.WriteLine(message);
  }

  // Función con varios parámetros
  public static void AddNumbers(int a, int b)
  {
    Console.WriteLine($"Suma: {a + b}");
  }

  // Función con retorno
  public static int MultiplyNumbers(int a, int b)
  {
    return a * b;
  }

  // Función externa con función interna (local)
  public static void OuterFunction()
  {
    // Función local dentro de una función
    static void InnerFunction()
    {
      Console.WriteLine("Soy una función interna.");
    }

    Console.WriteLine("Soy la función externa.");
    InnerFunction(); // Llamada a la función interna
  }

  // Modificar la variable global
  public static void ModifyGlobalVariable()
  {
    globalVariable += 10;
  }

  // Función que recibe dos cadenas de texto y retorna un número (Dificultad Extra)
  public static int PrintNumbersWithText(string text1, string text2)
  {
    int count = 0;

    for (int i = 1; i <= 100; i++)
    {
      if (i % 3 == 0 && i % 5 == 0)
      {
        // Si es múltiplo de 3 y 5, imprime ambas cadenas concatenadas
        Console.WriteLine(text1 + text2);
      }
      else if (i % 3 == 0)
      {
        // Si es múltiplo de 3, imprime la primera cadena
        Console.WriteLine(text1);
      }
      else if (i % 5 == 0)
      {
        // Si es múltiplo de 5, imprime la segunda cadena
        Console.WriteLine(text2);
      }
      else
      {
        // Imprime el número y aumenta el contador
        Console.WriteLine(i);
        count++;
      }
    }

    // Retorna la cantidad de veces que se imprimió un número
    return count;
  }
}
